### DISCUSSION TITLE
Rendering flash notifications

### DISCUSSION THREAD

**Discussion by gggiii** [livewire/livewire#6561] [1. Help] [answered]
Using livewire:3.0.1

I am trying to make a unified way of displaying flash messages as toast notifications in the base layout.

In the documentation ([link](https://livewire.laravel.com/docs/redirecting#flashing-without-redirecting)) it states to use
```php
session()->flash('success', 'Post updated!')
```
to set a flash message without having to redirect, however, these messages are accessible only in the component that triggers them.
Example: Add  `@dump(session('success'))` to:
- The root layout
- The page component
- A subcomponent(which calls the action)
the message is only displayed in the subcomponent.

With this behaviour, there needs to be a per-component implementation for rendering flash messages without a way to make it global (in the base layout or page component).

The only way to make flash messages accessible in the layout I found to be,
```php
session()->flash('success', 'Message');
return $this->redirect('/');
```
as the advertised
```php
return $this->redirect('/posts')->with('status', 'Post updated!');
```
doesn't work due to an error "Call to a member function with() on null", as described [here](https://github.com/livewire/livewire/discussions/6176)

**Comment by austenc** [comment, marked as answer]
You're right, using the `session()->flash()` method needs to be handled in your component. That's because when Livewire makes an ajax request, it's not re-rendering the entire layout each time. Rather, it only renders and injects the part of the DOM that is related to your component. 

One way to handle toast notifications is to dispatch an event from the Livewire component, and then handle that event within another component (with Alpine).

For example (in `AppServiceProvider`'s `boot()` method or a dedicated macro provider :
```
Component::macro('notify', function ($content, $title = null, $type = 'success') {
    $this->dispatch('notify', [
        'id' => uniqid(),
        'content' => $content,
        'title' => $title,
        'type' => $type,
    ]);
});
```

Then in your layout you can have an Alpine component that does something like:
```
<div x-data @notify.window="console.log($event.detail)">
```
This is a simplified example of the Alpine component. Instead of `console.log` you might have it call a function to display the toast message, or to put the event's details into a `messages` array on the javascript side. 

The nice thing about the component macro is using it becomes very simple:
```php
// In any Livewire component
$this->notify('Updated successfully');
```

**Reply by shaikhfoysal** [reply]
How to call it in javascript? Thanks

**Reply by austenc** [reply]
You can use `$wire` from within an alpine component to dispatch the event.

**Reply by shaikhfoysal** [reply]
Okay Thanks

**Reply by SagarNaliyapara** [reply]
The Livewire documentation on flash messages (https://livewire.laravel.com/docs/redirecting#flash-messages) may be missing some information. I've followed the instructions as outlined, but it's not working as expected. 

Thank you for your insights here @austenc something could be mentioned in documentation.

**Reply by hasan-aftab** [reply]
Thanks @austenc brother, I was using Trait to dispatch toast event, thanks for making me familiar with Macro.

**Comment by hasan-aftab** [comment]
create session in livewire from where u wanna dispatch toast:

        session(['verify-new-email' => 'success']);
 
 target livewire component where u wanna display the toast:
 
     public function mount()
    {
        $this->verifyNewEmailSession = session()->has('verify-new-email');
        $this->verifyNewEmailSessionMessage = session('verify-new-email');
        session()->forget('verify-new-email');
    }
    
create a dynamic toast event listener/dispatcher component
    
        <x-full-page-flash-message-event-js
        wire:key="verify-new-email"
        :flag="$verifyNewEmailSession"
        event="triggerVerifyNewEmailToastNotification"
        reset-session-event="resetVerifyNewEmailSession"
    /> 

dynamic prop values would be passed here:
    
it store the component values in the js variables then on livewire navigated event, passing component values to the dispatch method so could trigger livewire event.

```
@props([
    'flag' => false,                           // Boolean flag to decide whether to dispatch the event
    'event' => '',             // Name of the event to dispatch for flash messages
    'resetSessionEvent' => '', // Dynamic event for resetting the session
    'delay' => 100,                            // Delay in milliseconds before dispatching the flash event
])
<script>
    (function () {
        let flashFlag = @js($flag);
        let eventName = @js($event);
        let resetSessionEvent = @js($resetSessionEvent);
        let delay = @js($delay);

        document.addEventListener('livewire:navigated', function () {
            if (flashFlag) {
                setTimeout(() => Livewire.dispatch(eventName), delay);
                Livewire.dispatch(resetSessionEvent);
                flashFlag = false;
            }
        });
    })();
</script>

```
when event dispatches then, here its gonna be listen, which is triggering the toast event:

    #[On('triggerVerifyNewEmailToastNotification')]
    public function triggerVerifyNewEmailToastNotification()
    {
        $title = 'Invalid Link';
        $description = 'Please request a new verification link.';
        $type = 'danger';

        if ($this->verifyNewEmailSessionMessage === 'expired') $title = 'Link Expired';

        if ($this->verifyNewEmailSessionMessage === 'success') {
            $title = 'Email Updated';
            $description = 'Your email has been updated and verified.';
            $type = 'success';
        }

        $this->dispatchToastNotification($title, $description, $type);
    }

alongside the property and message values are being reset:

    #[On('resetVerifyNewEmailSession')]
    public function resetVerifyNewEmailSession()
    {
        $this->verifyNewEmailSession = false;
        $this->verifyNewEmailSessionMessage = '';
    }

**Comment by hasan-aftab** [comment]
here is the most refined and better solution:

create an inline livewire component, that listens to public session flash and then trigger its method which actually triggering the macro or trait method which gonna trigger toast.

add this component to layout so ur flash session can be listened from anywhere.

trigger flash message:

```
session()->flash('toast', [
            'title' => 'Email Verification Link Sent',
            'description' => 'Please check your email for a verification link.',
]);
```

```
<?php

namespace App\Livewire;

use Livewire\Component;

class ToastEventDispatch extends Component
{
    public function toastNotificationListener($title, $description, $type)
    {
        $this->dispatchToastNotification($title, $description, $type);
    }

    public function render()
    {
        return <<<'HTML'
        <div>
        @script
            <script>
                let toastData = @js(session('toast'));
                if (toastData) {
                    $wire.toastNotificationListener(toastData.title, toastData.description, toastData.type = 'success');
                }
            </script>
        @endscript
        </div>
        HTML;
    }
}

```

**Reply by joshhanley** [reply]
@hasan-aftab please stop yelling. The all caps is not needed.

**Reply by hasan-aftab** [reply]
@joshhanley mate, that was unintentional. I rolled back; never mind.

**Reply by joshhanley** [reply]
@hasan-aftab no worries! Thanks for updating 🙂

**Reply by hasan-aftab** [reply]
@joshhanley sent u  req on discord, would love to interact

**Reply by nizzam** [reply]
I didnt get your example. Could you please elaborate more.

**Reply by hasan-aftab** [reply]
@nizzam My apologies for the late response, I didn’t get a chance to check my emails earlier.

So here’s what I’ve implemented:

I created a macro service class that defines a dispatchToastNotification method. By using Laravel macros, this method is now globally available across the entire project.

The dispatchToastNotification method accepts parameters and dispatches a custom event to trigger toast notifications via a Blade component.

To handle notifications specifically during Livewire full-page navigation, I introduced a dedicated Livewire component class called ToastEventDispatch. The purpose of this class is to catch flash session data after a full-page component finishes loading and then emit the toast event.

Here's how it works:

When navigating between Livewire pages, I set a flash session with the toast data.

Once the new Livewire component is fully rendered, ToastEventDispatch checks for this session and triggers the appropriate notification.

This approach ensures that toast notifications appear after the component is fully mounted, resolving timing issues that typically occur with Livewire navigation.

Let me know if you’d like a code example or further explanation!
